### PR TITLE
Issue_1244: Saving settings in localStorage

### PR DESCRIPTION
**Pull request by arslanbekova**

### CODE REVIEW

File: src/features/filter/filter.selectors.ts [hunk 1]
```diff
@@ -22,7 +22,11 @@ export const { selectAll: selectFilters, selectById: selectFilterById } = filter
 export const selectClassFilterByPO = createSelector(
   (state: RootState, id?: PageObjectId) => {
     const pageObject = !isNil(id) ? selectPageObjById(state, id) : selectCurrentPageObject(state);
-    const filter = selectFilters(state).find(({ pageObjectId }: Filter) => pageObjectId === pageObject?.id);
+    const savedFilters = localStorage.getItem("filters");
```
Comment (MariiaNebesnova) [inline src/features/filter/filter.selectors.ts:25]: selectors is a thing only for getting data from state. Using them for something else potentially lead to bugs/ Selector is a memoized function under the hood, so we can face with case, when it doesn't retrieve an actual data from local storage.

File: src/features/filter/reducers/toggleClassFilterAll.thunk.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      ) => {
+        const { newValue, filter, library } = payload;
+
+        filterAdapter.upsertOne(state, { ...newValue, [FilterKey.JDIclassFilter]: filter });
+
+        if (!localStorage.getItem("filters")) {
+          localStorage.setItem("filters", JSON.stringify({ [library]: filter }));
+        } else {
+          const savedFilters = JSON.parse(localStorage.getItem("filters")!);
+          localStorage.setItem("filters", JSON.stringify({ ...savedFilters, [library]: filter }));
```
Comment (MariiaNebesnova) [inline src/features/filter/reducers/toggleClassFilterAll.thunk.ts:44]: the same

File: src/features/pageObjects/reducers/addPageObject.thunk.js [hunk 3]
```diff
@@ -10,13 +10,30 @@ import {
...
   const { title, url } = res[0].result;
   const className = getClassName(title);
-  const lastSelectedLibrary = selectLastElementLibrary(thunkAPI.getState());
-  const lastSelectedLocatorType = selectLastLocatorType(thunkAPI.getState());
+
+  let lastSelectedLibrary;
+  let lastSelectedLocatorType;
+
+  if (localStorage.getItem("library")) {
+    lastSelectedLibrary = localStorage.getItem("library");
```
Comment (MariiaNebesnova) [inline src/features/pageObjects/reducers/addPageObject.thunk.js:24]: what would happened, if we try to .getItem that doesn't exist in local storage ?

File: src/features/pageObjects/reducers/addPageObject.thunk.js [hunk 4]
```diff
@@ -10,13 +10,30 @@ import {
...
+  } else {
+    lastSelectedLibrary = selectLastElementLibrary(thunkAPI.getState()) || defaultLibrary;
+    localStorage.setItem("library", lastSelectedLibrary);
+  }
+
+  if (localStorage.getItem("locatorType")) {
```
Comment (MariiaNebesnova) [inline src/features/pageObjects/reducers/addPageObject.thunk.js:30]: also, please, save local storage keys to constants and use prefix `JDN_`

File: src/features/pageObjects/reducers/addPageObject.thunk.js (hunk 3 above)
Comment (arslanbekova) [reply src/features/pageObjects/reducers/addPageObject.thunk.js:24]: null will be returned

File: src/features/pageObjects/reducers/addPageObject.thunk.js [hunk 5]
```diff
@@ -48,11 +65,11 @@ export const addPageObjReducer = (builder) => {
         id,
         name,
         url,
-        library: lastSelectedLibrary || defaultLibrary,
+        library: lastSelectedLibrary,
         pathname,
         search,
         origin,
-        ...(lastSelectedLocatorType ? { locatorType: lastSelectedLocatorType } : {}),
+        locatorType: lastSelectedLocatorType,
```
Comment (MariiaNebesnova) [inline src/features/pageObjects/reducers/addPageObject.thunk.js:72]: according to our types, locatorType is optional, no need to store default value here

Comment (arslanbekova) [reply src/features/pageObjects/reducers/addPageObject.thunk.js]: fixed

File: src/features/pageObjects/reducers/addPageObject.thunk.js (hunk 4 above)
Comment (arslanbekova) [reply src/features/pageObjects/reducers/addPageObject.thunk.js]: done

File: src/features/filter/reducers/toggleClassFilterAll.thunk.ts (hunk 2 above)
Comment (arslanbekova) [reply src/features/filter/reducers/toggleClassFilterAll.thunk.ts]: fixed

File: src/features/filter/reducers/toggleClassFilter.thunk.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        initialFilter[jdiClass] = value;
+        filterAdapter.addOne(state, {
+          pageObjectId,
+          [FilterKey.JDIclassFilter]: { ...initialFilter },
+        });
+
+        if (!localStorage.getItem("filters")) {
+          localStorage.setItem("filters", JSON.stringify({ [library]: { ...initialFilter } }));
+        } else {
+          localStorage.setItem("filters", JSON.stringify({ ...savedFilters, [library]: initialFilter }));
```
[the remark this answers is not in the document]
Comment (arslanbekova) [reply src/features/filter/reducers/toggleClassFilter.thunk.ts]: fixed

File: src/features/filter/filter.selectors.ts (hunk 1 above)
Comment (arslanbekova) [reply src/features/filter/filter.selectors.ts]: fixed

File: src/features/filter/reducers/toggleClassFilter.thunk.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    const { pageObjectId, jdiClass, value, library } = payload;
+
+    const state = getState() as RootState;
+    const newFilterValue = simpleSelectFilterById(state.filters, pageObjectId);
+
+    const savedFilters = JSON.parse(localStorage.getItem(LocalStorageKey.Filter)!);
+    let newFilter;
+    const initialFilter = {
+      ...jdiClassFilterInit(library),
+      ...(savedFilters && savedFilters[library] && savedFilters[library]),
```
Comment (MariiaNebesnova) [inline src/features/filter/reducers/toggleClassFilter.thunk.ts:36]: why do you write `savedFilters[library]` twice?

Comment (arslanbekova) [reply src/features/filter/reducers/toggleClassFilter.thunk.ts:36]: first one is checking, that savedFilters has library property

File: src/features/pageObjects/reducers/addPageObject.thunk.js [hunk 8]
```diff
@@ -10,13 +10,33 @@ import {
...
+  const state = getState();
+
+  let lastSelectedLibrary;
+  let lastSelectedLocatorType;
+
+  if (localStorage.getItem(LocalStorageKey.Library)) {
+    lastSelectedLibrary = localStorage.getItem(LocalStorageKey.Library);
+  } else {
+    lastSelectedLibrary = selectLastElementLibrary(state) || defaultLibrary;
+    localStorage.setItem(LocalStorageKey.Library, lastSelectedLibrary);
```
Comment (MariiaNebesnova) [inline src/features/pageObjects/reducers/addPageObject.thunk.js:30]: why do you save the value at creating PO, I'm sure we need to save it only when user change it

File: src/features/pageObjects/reducers/addPageObject.thunk.js [hunk 9]
```diff
@@ -10,13 +10,33 @@ import {
...
+  }
+
+  if (localStorage.getItem(LocalStorageKey.LocatorType)) {
+    lastSelectedLocatorType = localStorage.getItem(LocalStorageKey.LocatorType);
+  } else {
+    lastSelectedLocatorType = selectLastLocatorType(state);
+    localStorage.setItem(LocalStorageKey.LocatorType, lastSelectedLocatorType ?? LocatorType.xPath);
+  }
```
Comment (MariiaNebesnova) [inline src/features/pageObjects/reducers/addPageObject.thunk.js:38]: same here

File: src/features/filter/reducers/toggleClassFilter.thunk.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+      ...jdiClassFilterInit(library),
+      ...(savedFilters?.[library] && savedFilters[library]),
+    };
+
+    if (newFilterValue) {
+      newFilter = { ...newFilterValue[FilterKey.JDIclassFilter], [jdiClass]: value };
+      setLocalStorage(LocalStorageKey.Filter, { ...savedFilters, [library]: newFilter });
+    } else {
+      initialFilter[jdiClass] = value;
+      if (!getLocalStorage(LocalStorageKey.Filter)) {
```
Comment (MariiaNebesnova) [inline src/features/filter/reducers/toggleClassFilter.thunk.ts:44]: why do you get it again? you have it on line 32

File: src/features/filter/reducers/toggleClassFilterAll.thunk.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    Object.keys(newFilter).forEach((key: string) => {
+      // don't know how to fix it
+      //@ts-ignore
+      newFilter[key] = value;
+    });
+
+    if (!getLocalStorage(LocalStorageKey.Filter)) {
+      setLocalStorage(LocalStorageKey.Filter, { [library]: newFilter });
+    } else {
+      const savedFilters = getLocalStorage(LocalStorageKey.Filter);
```
Comment (MariiaNebesnova) [inline src/features/filter/reducers/toggleClassFilterAll.thunk.ts:30]: instead of evaluate `getLocalStorage` twice, it's better to evaluate it before `if`

**Review by MariiaNebesnova — APPROVED** [review]
Great job! Thank you!

**Pull request merged**